### PR TITLE
[Transform] avoid mapping problems with index templates

### LINKED ISSUE #51321: [Transform] create explicit object mappings for nested field
When auto-creating the destination index, transform deducts appropriate mappings, however it creates the mappings only for the leafs, meaning if the field is nested like `a.b.c` it creates a mapping for `a.b.c` only. This opens up a corner case:

- a conflicting index template could exists for `a` or `a.b` (as reported by a user)

(Note that a conflicting mapping in the config is already handled in validation)

**Solution**: 

 - introduce explicit object mappings for nested fields

### PR DESCRIPTION
**Pull request by hendrikmuhs**
insert explict mappings for objects in nested output to avoid clashes with index templates

fixes #51321

### CODE REVIEW

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/SchemaUtil.java [hunk 1]
```diff
@@ -165,34 +171,75 @@ public static void getDestinationFieldMappings(final Client client,
...
+     * in which case it creates additional mappings:
+     *
+     * a.b : object
+     * a : object
+     *
+     * avoids snafu with index templates injecting incompatible mappings
+     *
+     * @param fieldMappings field mappings to inject to
+     */
+    static void insertNestedObjectMappings(Map<String, String> fieldMappings) {
```
Comment (hendrikmuhs) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/SchemaUtil.java]: essential change in this file, called on line 176

**Comment by hendrikmuhs** [comment]
run elasticsearch-ci/packaging-sample-matrix-unix

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/SchemaUtil.java [hunk 2]
```diff
@@ -165,34 +171,75 @@ public static void getDestinationFieldMappings(final Client client,
...
+        Map<String, String> additionalMappings = new HashMap<>();
+        fieldMappings.keySet().stream().filter(key -> key.contains(".")).forEach(key -> {
+            int pos;
+            String objectKey = key;
+            // lastIndexOf returns -1 on mismatch, but to disallow empty strings check for > 0
```
Comment (davidkyle) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/SchemaUtil.java]: An empty string would also return -1. Do you mean disallow strings containing a single '.' that is at the beginning? Wouldn't that be an error case anyway?

I ran a quick check in jshell:
```
jshell> "".lastIndexOf('.')
$3 ==> -1
```

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/SchemaUtil.java (hunk 1 above)
Comment (davidkyle) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/SchemaUtil.java]: 👍 thanks for differentiating the code changes from the formatting changes

**Review by davidkyle — APPROVED** [review]
LGTM. 

Left a comment about a comment

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/SchemaUtil.java (hunk 2 above)
Comment (hendrikmuhs) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/SchemaUtil.java]: yes, when I worked on this PR, I discovered that you could create fields with '.' at start or end, which did not work properly (e.g. bulk indexing error). I fixed it in #51369. 

So, you are right: this code is "defense in depth".

**Pull request merged**